scripts/git-purge-files

```diff
@@ -1,46 +1,142 @@
-#
-#    Remove all "*.bin" files from all directories:
-#    git-purge-path "\.bin$"
-#
-#    Remove all files in the "/foo" directory:
-#    git-purge-path "^/foo/$"
-#
-# Author: Lars Schneider, https://github.com/larsxschneider
+# Purge files from Git repositories
 #
 
+use 5.010;
 use strict;
 use warnings;
+use Getopt::Std;
+use File::Temp qw/ tempdir /;
+
+sub usage() {
+    print STDERR <<END;
+NAME
+       git-purge-files - Purge files from Git repositories
+
+SYNOPSIS
+       git-purge-files [-c] [-d] [-h] [<path-regex>] ...
+
+
+DESCRIPTION
+       This command purges files from a Git history by rewriting all
+       commits. Please note that this changes all commit hashes in the
+       history and therefore all branches and tags.
+
+       You want to run this script on a case sensitive file-system (e.g.
+       ext4 on Linux). Otherwise the resulting Git repository will not
+       contain changes that modify the casing of file paths.
+
+OPTIONS
+       <path-regex>...
+           A list of regular expression that defines what files should
+           be purged from the history. Use a `/` to anchor a path to the
+           root of the repository.
+
+       -c
+           Run in checking mode. The script will run the underlaying
+           `git fast-export | git fast-import` command without any
+           modifications to the data stream. Afterwards the input
+           repository is compared against the output repository.
+
+           ATTENTION: Although we run a check here, the repository
+                      under test is rewritten and potentially modified!
+
+       -d
+           Enable diff mode. This makes the underlaying `git fast-export`
+           output only the file differences between two commits. This
+           mode is quicker but more error prone. It is not recommended
+           in production usage.
+
+           See examples for potential problems here:
+           https://public-inbox.org/git/CABPp-BFLJ48BZ97Y9mr4i3q7HMqjq18cXMgSYdxqD1cMzH8Spg\@mail.gmail.com/
+
+       -h
+           This help.
 
-my $path_regex = join( "|", @ARGV );
+EXAMPLES
+       o   Remove the file "test.bin" from all directories:
 
-open( my $pipe_in, "git fast-export --progress=10000 --no-data --all --signed-tags=warn-strip --tag-of-filtered-object=rewrite |" ) or die $!;
-open( my $pipe_out, "| git fast-import --force --quiet" ) or die $!;
+               \$ git-purge-path "/test.bin$"
+
+       o   Remove all "*.bin" files from all directories:
+
+               \$ git-purge-path "\.bin$"
+
+       o   Remove all files in the "/foo" directory:
+
+               \$ git-purge-path "^/foo/$"
+END
+    exit(1);
+}
+
+our($opt_h, $opt_d, $opt_c);
+getopts("hdc") or usage();
+usage if $opt_h;
+
+# TODO: Git 2.23 will likely have a "--reencode=no" option that we want add here
+my $export_opts = "--all --no-data --progress=1000 --signed-tags=warn-strip --tag-of-filtered-object=rewrite --use-done-feature";
```

Comment:
@larsxschneider Seems like [2.23](https://github.com/git/git/blob/master/Documentation/RelNotes/2.23.0.txt) is out now and `--reencode=no` is added: https://github.com/git/git/compare/v2.22.0...v2.23.0#diff-3425c1401396aa19fb95ae026d449429